thrift/shared.thrift

```diff
@@ -56,6 +56,10 @@ exception ServiceBusyError {
   1: required string message
 }
 
+exception UserServiceBusyError {
```

Comment:
I haven't thought about it deeply, but I'd leant to I'd suggest that too. 

There's two bit's I'm not sure about: 

Firstly, I'm also not sure that rate-limit errors are not retriable? Even for the hot-shard rate-limiter, but rather they just need to ensure callers retry with backoff. I'm assuming you're referring to the user-facing context, but even there (say a worker is returning a decision task as being complete), you probably always want to encourage backoff and retry. 

To the second point, introducing a new type will mean a fair amount of work to ensure this new error is handled in the same way as the other service-busy errors (if it's a plain error you'll probably just get retry-storms), so you'll have to add any new error you create into quite a few places to ensure it's handled correctly.

Reply:
The idea of this is:

The new error will only be used between cadence-frontend and cadence-history.

Currently when History returns a ServiceBusyError Frontend will retry, the only thing clients see is that requests are suddenly much slower (I personally spend like a day trying to understand why they were suddenly so slow, when I made the integration test).

The new logic is, if history returns a ServiceBusyError frontend will do retires (as before), if history returns a UserServiceBusyError, frontend will immediately return a ServiceBusyError (not a UserServiceBusyError).

We don't want to handle this internally in the server as we want the users to see they are doing something wrong.
For this error (wfID ratelimiting) it's clear which user is doing something bad, so we can return that information to them. (for e.g. the persistence rate limit it's much less clear who is causing it as it is cluster global, so it makes more sense to handle internally).

We could add a field to ServiceBusyError which can tell frontend if it should retry or just return immediately, with the same effect.
I'd lean to having a new internal error type to avoid exposing internal details to the users.

Happy to discuss 🙂

Reply:
I guess there are 3 options we can narrow down to
1. Adding a `Reason` field to `ServiceBusyError` and building logic around that in Frontend . Frontend owns retry or not decision in this case
2. Adding a `Retriable bool` field `ServiceBusyError`. History owns retry or not decision in this case.
3. Adding a new type `UserServiceBusyError` (or rename to something even more specific). Frontend owns retry or not decision in this case.

I'd go with 1 or maybe 2 if we want to shift responsibility to to history. 
Let's hear feedback from @Shaddoll and @Groxx as well

Reply:
It makes sense to not retry `frontend->history` flow in this case. Maybe we should consider removing `frontend -> history` retries on all `ServiceBusyError`. The retries at multiple layers can easily create a retry storm.

Reply:
From past struggles around retries I'd probably say:
- a unique top-level type per fine-grained error-handling-strategy is just too much overhead, so some kind of common top level type and enum-y behavior often makes sense.
- "reason" does not define fallback behavior when a reason is not recognized
  - particularly relevant when deploying upgrades
  - this can always become a 500-server-error, but what kind of retries does that trigger?  retry-behavior is still being decided, even if it's not in our codebase.  and 500s tend to mean "yes retry".
- "retriable bool" sounds good but falls apart in situations like this where you want one layer to not-retry and another layer to yes-retry
  - but it does solve the fallback issue

Which kinda leaves things in a nasty corner where you probably, reasonably, want both reason _and_ retriable, but it's unclear what you should do if they disagree somehow.

---

I think part of the problem stems from whether an error is expected to be _handled internally and immediately_, or bubbled implicitly to be given some kind of fallback behavior somewhere else.

Unhandled errors always bubble because they essentially must (with only rare exceptions), other types... not sure.  To some degree I would love to say "each service should have its own errors, and force conversion when crossing boundaries" because that'd make it clear (nothing bubbles across RPC, each decides based on the thing it came from) but that'd be a big change.

Reply:
ignoring that rambling: I do think that both "message" and "cause" fields should be standard.  otherwise chaining cause-of-error is next to impossible.

Reply:
Not as part of this PR but we may consider standardizing our RPC error status codes with [this one](https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto#L25) which is wrapped inside [this](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto). 

For now introducing a `reason` field and only not retrying for deny-listed reasons will be most backward compatible path forward. To achieve what @jakobht is trying to do we just define `reason=workflowid_ratelimited` and stop retrying for that. 
Then as a follow up I'm in favor of not retrying at all for `ServiceBusyError` or change the backoff/timeout parameters to avoid making situations worse.

Reply:
Sounds good, will do this change, thanks all for the input!

I agree it would be good to take a more general look at the error handling

Reply:
yeah, the google/rpc/status is very much a standard, worth using.
if we wanted to stick to it strictly, that'd mean that "reason" goes into one of the `Any` fields (we can have a reason-type)